tool/tsh/gcp.go

```diff
@@ -0,0 +1,443 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"crypto/tls"
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+	"sort"
+	"strings"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/asciitable"
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/srv/alpnproxy"
+	alpncommon "github.com/gravitational/teleport/lib/srv/alpnproxy/common"
+	"github.com/gravitational/teleport/lib/tlsca"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+const (
+	gcloudCLIBinaryName = "gcloud"
+)
+
+func onGCP(cf *CLIConf) error {
+	app, err := pickActiveGCPApp(cf)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = app.StartLocalProxies()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	defer func() {
+		if err := app.Close(); err != nil {
+			log.WithError(err).Error("Failed to close GCP app.")
+		}
+	}()
+
+	args := cf.GCPCommandArgs
+
+	cmd := exec.Command(gcloudCLIBinaryName, args...)
+	return app.RunCommand(cmd)
+}
+
+// gcpApp is an GCP app that can start local proxies to serve GCP APIs.
+type gcpApp struct {
+	cf      *CLIConf
+	profile *client.ProfileStatus
+	app     tlsca.RouteToApp
+	secret  string
+
+	localALPNProxy    *alpnproxy.LocalProxy
+	localForwardProxy *alpnproxy.ForwardProxy
+}
+
+// newGCPApp creates a new GCP app.
+func newGCPApp(cf *CLIConf, profile *client.ProfileStatus, app tlsca.RouteToApp) (*gcpApp, error) {
+	secret, err := getGCPSecret()
+	if err != nil {
+		return nil, err
+	}
+	return &gcpApp{
+		cf:      cf,
+		profile: profile,
+		app:     app,
+		secret:  secret,
+	}, nil
+}
+
+// getGCPSecret will return fresh secret to use or read it from environment.
+func getGCPSecret() (string, error) {
+	secret := os.Getenv("TELEPORT_GCP_SECRET")
```

Comment:
I think we should make this a constant in constants.go with other env vars we define

Reply:
`tsh` has a bunch of env vars defined for its own use only, it is rightfully in `tsh.go`. I'll add an entry there.